File: class/class9.1/MyDataType.cpp
#include <iostream>
#include <vector>
#include <string>
#include <utility>
#include <variant>
#include <fstream>
#include <cstdint>

class MyDataType {
public:
    using DataType = std::variant<int16_t, std::string, std::vector<std::pair<std::string, MyDataType>>>;

    MyDataType(int16_t value) : data(value) {}
    MyDataType(const std::string& value) : data(value) {}
    MyDataType(const std::vector<std::pair<std::string, MyDataType>>& value) : data(value) {}

    const DataType& getData() const {
        return data;
    }

private:
    DataType data;
};


class MsgPackSerializer {
public:
    void serialize(const MyDataType& obj, std::ostream& out) {
        serializeData(obj.getData(), out);
    }

private:
    void serializeData(const MyDataType::DataType& data, std::ostream& out) {
        std::visit([&](const auto& value) {
            using T = std::decay_t<decltype(value)>;
            if constexpr (std::is_same_v<T, int16_t>) {
                serializeInt(value, out);
            } else if constexpr (std::is_same_v<T, std::string>) {
                serializeString(value, out);
            } else if constexpr (std::is_same_v<T, std::vector<std::pair<std::string, MyDataType>>>) {
                serializeMap(value, out);
            }
        }, data);
    }

    void serializeInt(int16_t value, std::ostream& out) {
        out.put(0xd1);  
        out.put(static_cast<char>((value >> 8) & 0xFF));
        out.put(static_cast<char>(value & 0xFF));
    }

    void serializeString(const std::string& value, std::ostream& out) {
        out.put(0xd9); 
        out.put(static_cast<uint8_t>(value.size()));
        out.write(value.data(), value.size());
    }

    void serializeMap(const std::vector<std::pair<std::string, MyDataType>>& map, std::ostream& out) {
        out.put(0xde);  
        out.put(static_cast<char>((map.size() >> 8) & 0xFF));
        out.put(static_cast<char>(map.size() & 0xFF));
        for (const auto& [key, val] : map) {
            serializeString(key, out);
            serializeData(val.getData(), out);
        }
    }
};

int main() {
   
    MyDataType intData(42);
    MyDataType stringData(std::string("Hello, MsgPack"));
    std::vector<std::pair<std::string, MyDataType>> mapData = {
        {"key1", MyDataType(123)},
        {"key2", MyDataType("value2")}
    };
    MyDataType mapDataType(mapData);

    MsgPackSerializer serializer;
    std::ofstream outFile("output.msgpack", std::ios::binary);

    serializer.serialize(intData, outFile);
    serializer.serialize(stringData, outFile);
    serializer.serialize(mapDataType, outFile);

    outFile.close();
    return 0;
}

File: home/home9.1/src/main/java/com/example/MyDataType.java
package com.example;

import org.msgpack.core.MessageBufferPacker;
import org.msgpack.core.MessagePack;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class MyDataType {
    private final Object data;

    public MyDataType(int value) {
        this.data = value;
    }

    public MyDataType(String value) {
        this.data = value;
    }

    public MyDataType(List<Map.Entry<String, MyDataType>> value) {
        this.data = value;
    }

    public Object getData() {
        return data;
    }

    public static void serialize(MyDataType obj, OutputStream out) throws IOException {
        MessageBufferPacker packer = MessagePack.newDefaultBufferPacker();
        serializeData(obj.getData(), packer);
        packer.close();
        out.write(packer.toByteArray());
    }

    private static void serializeData(Object data, MessageBufferPacker packer) throws IOException {
        if (data instanceof Integer) {
            packer.packInt((int) data);
        } else if (data instanceof String) {
            packer.packString((String) data);
        } else if (data instanceof List) {
            List<Map.Entry<String, MyDataType>> list = (List<Map.Entry<String, MyDataType>>) data;
            packer.packMapHeader(list.size());
            for (Map.Entry<String, MyDataType> entry : list) {
                packer.packString(entry.getKey());
                serializeData(entry.getValue().getData(), packer);
            }
        }
    }

    public static void main(String[] args) {
        try {
            MyDataType intData = new MyDataType(42);
            MyDataType stringData = new MyDataType("Hello, MsgPack");
            List<Map.Entry<String, MyDataType>> mapData = new ArrayList<>();
            mapData.add(Map.entry("key1", new MyDataType(123)));
            mapData.add(Map.entry("key2", new MyDataType("value2")));
            MyDataType mapDataType = new MyDataType(mapData);

            try (FileOutputStream out = new FileOutputStream("output.msgpack")) {
                serialize(intData, out);
                serialize(stringData, out);
                serialize(mapDataType, out);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
